runtime/lua/helpers/plugin/init.lua

```diff
@@ -0,0 +1,20 @@
+-- luacheck: globals unpack vim.api
+local nvim = vim.api
+local map = require('helpers.plugin.map').map
+local unmap = require('helpers.plugin.map').unmap
+local functions = require('helpers.plugin.map').functions
+
+local function new_plugin(name)
+  -- TODO(KillTheMule): Check assumptions about subsequent calls of this
+  local ns = nvim.nvim_create_namespace(name)
+
+  -- Should be redundant after the comment above has ben ascertained
+  assert(functions[ns] == nil, "Namspace "..tostring(ns).." already exists")
+
+  functions[ns] = {}
+  return { ns = ns, name = name, map = map, unmap = unmap }
```

Comment:
why does it return map/unmap?

In its current form `new_plugin` seems unnecessary. Instead "namespaces" could be the primary concept, and `map()`/`unmap()` could take `ns` as their first parameter. We can add `vim.create_namespace`.

Also unsure about adding stateful "plugin objects" which would need to be kept by plugins instead of just working with a namespace id.

Reply:
Well, it's exactly the statefulness that seemed useful to me, and which I wanted to discuss :)

Right now, this doesn't do a lot, true: It feels a tad more elegant to use imho, and I use the `ns` transparently for unmap. But having this plugin object would allow us to keep state for plugins, and provide features based on that. E.g. aucmd handling could be rather simplified by transparently defining them in groups, and providing a command to delete them on a fine-grained level. This could be extende up to providing a `plugin:unload` command that reverts everything the plugin has set, or a `plugin:reload` that does unloading+loading.

All this might be possible without such a plugin object, I guess, but it already feels a bit better for `map`, as I said. Maybe we could try and push more state into the plugin object itself, rather than the `map` module.

Reply:
It makes sense, but the disconnection from core could be confusing. E.g. if regular `:nmap` is used instead of `map()`, now `unmap()` can't be used. And even if one dutifully uses `map()` for everything, it will still override maps added by regular VimL plugins.  

What about a map "stack" for the each buffer (as well as global)? This would involve some changes to core, but perhaps not too drastic.

Reply:
Hmm, I'd not guess people would use viml `:map` from lua when they can use `plugin:map`, and the difference seemed easy to understand... override could be handled, we can check for existing maps.

Sure, working in core unifies things, but it's much harder, and in my head we don't only want this for maps, but rather aucmds, settings, definition, etc(?) as well, so we're looking at quite a stack of work, which is much easier to do in lua.

Reply:
What does "stack" mean here? Is it different/better than associating each map with the ns id?

Reply:
I assumed it was wrt to mapping certain keys several times. You can push the mappings on the stack, the last one is the active one, and if it gets popped, the next one is active.

Reply:
> not guess people would use viml :map from lua

For many filetypes there will be VimL plugins adding maps. So even if Lua plugin is well-behaved and uses only `map()`, this question will be raised by anyone writing a Lua plugin. So it weakens the utility of `map()`.

> Override could be handled, we can check for existing maps.

We can check, but what will we do? Without a "stack" we can only override or error.

> and in my head we don't only want this for maps, but rather aucmds, settings, definition, etc(?) as well, so we're looking at quite a stack of work, which is much easier to do in lua.

We will avoid a stack of work :) I'm only interested in simple, reliable approaches. 

But we should not add Lua mechanisms if they will commonly conflict with VimL mechanisms, that won't be useful and it will just add bloat. So we need to think about a compromise that ensures the Lua mechanisms will be useful.

[maktaba](https://github.com/google/vim-maktaba) took the approach that I'm trying to avoid. There's a reason no one uses maktaba even though it's a high quality project.

Reply:
> What does "stack" mean here? Is it different/better than associating each map with the ns id?

Yes, what @KillTheMule said. I'm not really fond of the idea because it seems over-engineered, but I suggested it as a way to let `map()`/`unmap()` work harmoniously with legacy VimL plugins. If we don't have a solution to that, then `map()`/`unmap()` are living in their own world and I suspect they will be ignored by plugin authors because of the impedance mismatch.

Thoughts?

Reply:
> For many filetypes there will be VimL plugins adding maps

Yeah, but they're OTHER plugnis, so `plugin:unmap` should not work on them (at least I did note intend it to :)).

> But we should not add Lua mechanisms if they will commonly conflict with VimL mechanism

What conflict are you thinking about, actually? As mentioned above, `plugin:unmap` is made so you can only unmap mappings made from the plugin itself, so that's the same for lua and viml (the only disconnect I see is when a plugin contains both viml and lua code).

> We can check, but what will we do? Without a "stack" we can only override or error.

My favourite way would be to return a table `maperror` that has info in why it failed (say, the mappings that already exist), and has a `force` method, so you can either do `if plugin:map() ~= {}` and handle the error yourself, or just go `if plugin:map().force()` to force your mapping.

> ... and I suspect they will be ignored by plugin authors because of the impedance mismatch.

Of course top priority is to make something people _want_ to use. But otoh viml has quite some snags, which we might be able to avoid. One huge point in our favour is that there's one less use-case for the lua api: I does not really need to be typeable very well on the command line :)

Reply:
> What conflict are you thinking about, actually? As mentioned above, plugin:unmap is made so you can only unmap mappings made from the plugin itself

The conflict is `map()`, not `unmap()`. `map()` will override maps from VimL plugins. So a Lua plugin will use `map()` but it still must deal with the wild west of VimL ...

> `plugin:map().force()`

This kind of jQuery-like API looks cool, but we should optimize for "obvious" and "easy to debug". `map('force', ...)` might better satisfy those guidelines.

Reply:
There is also the aspect that good plugins are not supposed to map a huge number of user accessible keys in the global namespace anyway. A good plugin should first describe its global actions with `<Plug>(MyPlugin-Action)` maps, so the mappings defined to lua functions should primarily be on this form. And then it could possibly provide an opt-in set of default mappings, defined as ordinary expansions to the `<Plug>` form. (There are of course some exceptions, like in a buffer owned by a plugin, it can go crazy with buffer-local mappings, and ftplugins using `<LocalLeader>` prefixed mappings)

-